File: itx-go/itx-backend/internal/state/state.go
package state

import (
	"fmt"

	"github.com/chehsunliu/itx/itx-go/itx-contract/queue"
	queuefactory "github.com/chehsunliu/itx/itx-go/itx-contract/queue/factory"
	"github.com/chehsunliu/itx/itx-go/itx-contract/repo/factory"
	"github.com/chehsunliu/itx/itx-go/itx-contract/repo/post"
	"github.com/chehsunliu/itx/itx-go/itx-contract/repo/subscription"
	"github.com/chehsunliu/itx/itx-go/itx-contract/repo/user"
	queuerabbit "github.com/chehsunliu/itx/itx-go/itx-impl/queue/rabbitmq"
	queuesqs "github.com/chehsunliu/itx/itx-go/itx-impl/queue/sqs"
	"github.com/chehsunliu/itx/itx-go/itx-impl/repo/mariadb"
	"github.com/chehsunliu/itx/itx-go/itx-impl/repo/postgres"
)

type AppState struct {
	Props                AppProps
	PostRepo             post.Repo
	UserRepo             user.Repo
	SubscriptionRepo     subscription.Repo
	ControlStandardQueue queue.MessageQueue
}

func FromEnv() (AppState, error) {
	props, err := PropsFromEnv()
	if err != nil {
		return AppState{}, err
	}

	dbProvider := props.DBProvider
	if dbProvider == "" {
		dbProvider = "postgres"
	}
	var repoFactory factory.RepoFactory
	switch dbProvider {
	case "postgres":
		f, err := postgres.New(props.Postgres)
		if err != nil {
			return AppState{}, err
		}
		repoFactory = f
	case "mariadb":
		f, err := mariadb.New(props.MariaDB)
		if err != nil {
			return AppState{}, err
		}
		repoFactory = f
	default:
		return AppState{}, fmt.Errorf("unknown ITX_DB_PROVIDER: %s", dbProvider)
	}

	queueProvider := props.QueueProvider
	if queueProvider == "" {
		queueProvider = "sqs"
	}
	var queueFactory queuefactory.MessageQueueFactory
	switch queueProvider {
	case "sqs":
		f, err := queuesqs.New(props.SQS)
		if err != nil {
			return AppState{}, err
		}
		queueFactory = f
	case "rabbitmq":
		f, err := queuerabbit.New(props.RabbitMQ)
		if err != nil {
			return AppState{}, err
		}
		queueFactory = f
	default:
		return AppState{}, fmt.Errorf("unknown ITX_QUEUE_PROVIDER: %s", queueProvider)
	}

	return AppState{
		Props:                props,
		PostRepo:             repoFactory.CreatePostRepo(),
		UserRepo:             repoFactory.CreateUserRepo(),
		SubscriptionRepo:     repoFactory.CreateSubscriptionRepo(),
		ControlStandardQueue: queueFactory.CreateControlStandardQueue(),
	}, nil
}

File: itx-rs/itx-backend/src/state.rs
use axum::extract::FromRef;
use itx_contract::repo::factory::RepoFactory;
use itx_contract::repo::post::PostRepo;
use itx_contract::repo::subscription::SubscriptionRepo;
use itx_contract::repo::user::UserRepo;
use itx_impl::repo::mariadb::MariaDbRepoFactory;
use itx_impl::repo::postgres::PostgresRepoFactory;
use std::error::Error;
use std::sync::Arc;

#[derive(Clone)]
pub struct AppState {
    pub post_repo: Arc<dyn PostRepo>,
    pub user_repo: Arc<dyn UserRepo>,
    pub subscription_repo: Arc<dyn SubscriptionRepo>,
}

impl AppState {
    pub async fn from_env() -> Result<Self, Box<dyn Error>> {
        let repo_factory: Arc<dyn RepoFactory> = match std::env::var("ITX_DB_PROVIDER").as_deref().unwrap_or("postgres")
        {
            "postgres" => Arc::new(PostgresRepoFactory::from_env().await?),
            "mariadb" => Arc::new(MariaDbRepoFactory::from_env().await?),
            other => panic!("unknown ITX_DB_PROVIDER: {other}"),
        };
        let post_repo = repo_factory.create_post_repo();
        let user_repo = repo_factory.create_user_repo();
        let subscription_repo = repo_factory.create_subscription_repo();

        Ok(Self {
            post_repo,
            user_repo,
            subscription_repo,
        })
    }
}

impl FromRef<AppState> for Arc<dyn PostRepo> {
    fn from_ref(app_state: &AppState) -> Self {
        app_state.post_repo.clone()
    }
}

impl FromRef<AppState> for Arc<dyn UserRepo> {
    fn from_ref(app_state: &AppState) -> Self {
        app_state.user_repo.clone()
    }
}

impl FromRef<AppState> for Arc<dyn SubscriptionRepo> {
    fn from_ref(app_state: &AppState) -> Self {
        app_state.subscription_repo.clone()
    }
}
